File: AlgoAnalysis-C/DynamicProgramming/LongestCommonSubsequenceProblem.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define MAX_LEN 100 // Tabloların maksimum boyutu

// Global Tablolar
// c: Uzunlukları tutar
// b: Yönleri tutar ('D': Diagonal/Çapraz, 'U': Up/Yukarı, 'L': Left/Sol)
int c[MAX_LEN][MAX_LEN];
char b[MAX_LEN][MAX_LEN];

// Prototipler
void LCS_LENGTH(char X[], char Y[]);
void PRINT_LCS(char X[], int i, int j);

int main() {
    // Slayt Sayfa 19'daki Örnek DNA Dizileri
    char X[] = "ABCBDAB";
    char Y[] = "BDCABA";

    printf("Dizi X: %s\n", X);
    printf("Dizi Y: %s\n", Y);

    // 1. Tabloları Doldur
    LCS_LENGTH(X, Y);

    // 2. Sonucu Yazdır
    printf("\nEn Uzun Ortak Altdizi (LCS): ");
    PRINT_LCS(X, strlen(X), strlen(Y)); // Sondan başa doğru yazdırır

    printf("\nLCS Uzunlugu (c tablosundan): %d\n", c[strlen(X)][strlen(Y)]);

    return 0;
}

// --- MOTOR KISMI: Tabloları Doldurma ---
void LCS_LENGTH(char X[], char Y[]) {
    // 1. Stringlerin gerçek uzunluklarını al
    int m = strlen(X);
    int n = strlen(Y);

    // 2. Tablonun "Duvarlarını" Ör (Initialization)
    // Boş bir string ile herhangi bir stringin LCS'si 0'dır.
    // Bu yüzden 0. satır ve 0. sütun tamamen 0 yapılır.
    for (int i = 0; i <= m; i++) {
        c[i][0] = 0;
    }
    for (int i = 0; i <= n; i++) {
        c[0][i] = 0;
    }

    // 3. Ana Döngü: Tabloyu doldurmaya başla
    // i: 1'den m'ye (Tablo satırları)
    // j: 1'den n'ye (Tablo sütunları)
    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {

            // --- KRİTİK NOKTA (BİT YENİĞİ) ---
            // Tabloda 'i'nci sıradayız ama String'de 'i-1'e bakmalıyız.
            // Çünkü Tablo 1'den, String 0'dan başlar.
            if (X[i - 1] == Y[j - 1]) {
                // EŞLEŞME VAR! (Harfler aynı)
                // Sol-Üst köşedeki değere 1 ekle.
                c[i][j] = c[i - 1][j - 1] + 1;
                b[i][j] = 'D'; // Yön: Diagonal (Çapraz/Sol-Üst)
            }
            // EŞLEŞME YOKSA: En büyük komşuyu seç
            else if (c[i - 1][j] >= c[i][j - 1]) {
                // Üstteki değer daha büyük (veya eşit)
                c[i][j] = c[i - 1][j];
                b[i][j] = 'U'; // Yön: Up (Yukarı)
            }
            else {
                // Soldaki değer daha büyük
                c[i][j] = c[i][j - 1];
                b[i][j] = 'L'; // Yön: Left (Sol)
            }
        }
    }
}

// --- RAPORTÖR KISMI: Sonucu Yazdırma ---
void PRINT_LCS(char X[], int i, int j) {
    // 1. Taban Durumu (Base Case)
    // Duvara (0. satır veya 0. sütun) çarparsak dur.
    if (i == 0 || j == 0) {
        return;
    }

    // 2. Yönleri Takip Et (Backtracking)
    if (b[i][j] == 'D') { // Eğer yön Çapraz ise...
        // Bu harf (X[i-1]) ortak harftir!

        // ÖNCE: Daha önceki harfleri bulmak için geriye (çapraz) git.
        PRINT_LCS(X, i - 1, j - 1);

        // SONRA: Harfi yazdır.
        // (Bu sıra sayesinde tersten değil, düzgün sırada yazarız)
        printf("%c", X[i - 1]);
    }
    else if (b[i][j] == 'U') { // Eğer yön Yukarı ise...
        // Bu harf eşleşmedi, yukarı git.
        PRINT_LCS(X, i - 1, j);
    }
    else if (b[i][j] == 'L'){ // Eğer yön Sol ise...
        // Bu harf eşleşmedi, sola git.
        PRINT_LCS(X, i, j - 1);
    }
}

File: AlgoAnalysis-Python/DynamicProgramming/LongestCommonSubsequenceProblem.py
# Maksimum dizi boyutu sabiti (C tarzı statik bellek yönetimi simülasyonu)
MAX_LEN = 100

# c tablosu: LCS uzunluklarını saklar. c[i][j], X'in ilk i karakteri ile
# Y'nin ilk j karakteri arasındaki LCS uzunluğudur.
c = [[0] * MAX_LEN for _ in range(MAX_LEN)]

# b tablosu: Yön bilgilerini (okları) saklar. Çözümü geriye doğru izlemek için kullanılır.
# 'D' (Diagonal/Çapraz), 'U' (Up/Yukarı), 'L' (Left/Sol) değerlerini alır.
b = [[0] * MAX_LEN for _ in range(MAX_LEN)]


# =============================================================================
# FONKSİYON: LCS_LENGTH
# AMAÇ: X ve Y dizilerinin LCS uzunluğunu hesaplar ve tabloları doldurur.
# =============================================================================
def LCS_LENGTH(X, Y):
    m = len(X)  # X dizisinin uzunluğu
    n = len(Y)  # Y dizisinin uzunluğu

    # 1. ADIM: Tablonun ilk sütununu 0 yap (Y boş küme ise LCS 0'dır).
    # range(0, m + 1) -> 0'dan m'e kadar.
    for i in range(0, m + 1, 1):
        c[i][0] = 0

    # 2. ADIM: Tablonun ilk satırını 0 yap (X boş küme ise LCS 0'dır).
    # range(0, n + 1) -> 0'dan n'e kadar.
    for j in range(0, n + 1, 1):
        c[0][j] = 0

    # 3. ADIM: Tabloları doldurma (Ana Döngü)
    # i, X dizisindeki karakterleri; j, Y dizisindeki karakterleri temsil eder.
    # Dikkat: Tablo indeksleri 1'den başlar, dizi indeksleri 0'dan.
    # Bu yüzden c[i][j] için X[i-1] ve Y[j-1]'e bakarız.
    for i in range(1, m + 1, 1):
        for j in range(1, n + 1, 1):

            # DURUM 1: Karakterler Eşleşirse (X[i-1] == Y[j-1])
            if X[i - 1] == Y[j - 1]:
                # Sol üst çaprazdaki (i-1, j-1) değeri 1 artır.
                # Çünkü bu karakteri LCS'ye ekledik.
                c[i][j] = c[i - 1][j - 1] + 1
                b[i][j] = 'D'  # Yön: Diagonal (Sol Üst Çapraz)

            # DURUM 2: Eşleşme Yoksa ve Üstteki Değer Soldakinden Büyük veya Eşitse
            elif c[i - 1][j] >= c[i][j - 1]:
                # Üstteki değeri aynen al (uzunluk değişmez).
                c[i][j] = c[i - 1][j]
                b[i][j] = 'U'  # Yön: Up (Yukarı)

            # DURUM 3: Eşleşme Yoksa ve Soldaki Değer Daha Büyükse
            else:
                # Soldaki değeri aynen al.
                c[i][j] = c[i][j - 1]
                b[i][j] = 'L'  # Yön: Left (Sol)


# =============================================================================
# FONKSİYON: PRINT_LCS
# AMAÇ: b tablosundaki yönleri takip ederek LCS'yi ekrana yazdırır.
# YÖNTEM: Geri izleme (Backtracking) ve Recursion (Özyineleme).
# =============================================================================
def PRINT_LCS(X, i, j):
    # Temel Durum: Tablonun kenarına (0. satır veya 0. sütun) ulaştıysak dur.
    if i == 0 or j == 0:
        return

    # Eğer yön 'D' (Diagonal) ise:
    # Bu, X[i-1] ve Y[j-1]'in eşleştiği ve LCS'nin bir parçası olduğu anlamına gelir.
    if b[i][j] == 'D':
        # Önce özyinelemeli olarak daha öncesini yazdır (Sıralı çıktı için).
        PRINT_LCS(X, i - 1, j - 1)
        # Dönüşte karakteri yazdır.
        print(f"{X[i - 1]}")

    # Eğer yön 'U' (Yukarı) ise:
    # Bu karakter eşleşmedi, yukarıdaki hücreye git.
    elif b[i][j] == 'U':
        PRINT_LCS(X, i - 1, j)

    # Eğer yön 'L' (Sol) ise:
    # Bu karakter eşleşmedi, soldaki hücreye git.
    elif b[i][j] == 'L':
        PRINT_LCS(X, i, j - 1)


# =============================================================================
# ANA BLOK (MAIN)
# =============================================================================
if __name__ == "__main__":
    # Test Dizileri
    # X = "ABCBDAB"
    X = ['A', 'B', 'C', 'B', 'D', 'A', 'B']
    # Y = "BDCABA"
    Y = ['B', 'D', 'C', 'A', 'B', 'A']

    print(f"Dizi X: {X}")
    print(f"Dizi Y: {Y}")

    # Uzunlukları hesapla ve tabloları doldur
    LCS_LENGTH(X, Y)

    print("\n En Uzun Ortak Altdizi (LCS):")
    # PRINT_LCS fonksiyonunu dizilerin son indeksinden başlatıyoruz.
    PRINT_LCS(X, len(X), len(Y))

    # c tablosunun sağ alt köşesi (c[m][n]) her zaman LCS'nin uzunluğunu verir.
    print(f"\nLCS Uzunluğu (c tablosundan): {c[len(X)][len(Y)]}")